graphql.go

```diff
@@ -307,24 +313,65 @@ func (c *Client) runWithJSON(ctx context.Context, req *Request, resp interface{}
 	}
 	c.logf(">> headers: %v", r.Header)
 	r = r.WithContext(ctx)
-	res, err := c.sendRequest(r)
-	if err != nil {
-		return err
-	}
-	defer res.Body.Close()
-	var buf bytes.Buffer
-	if _, err := io.Copy(&buf, res.Body); err != nil {
-		return errors.Wrap(err, "reading body")
-	}
-	c.logf("<< %s", buf.String())
-	if err := json.NewDecoder(&buf).Decode(&gr); err != nil {
-		return errors.Wrap(err, "decoding response")
-	}
-	if len(gr.Errors) > 0 {
-		// return first error
-		return gr.Errors[0]
+
+	return c.executeRequest(gr, r)
+}
+
+func (c *Client) executeRequest(gr *graphResponse, r *http.Request) error {
+	gqlRetryConfig := c.retryConfig
+	var body io.Reader = r.Body
+	tryCount := 0
+	for ; tryCount < gqlRetryConfig.MaxTries; tryCount++ {
+		buf := new(bytes.Buffer)
+		r.Body = ioutil.NopCloser(io.TeeReader(body, buf))
+
+		retryConfig := c.retryConfig
+		res, err := c.sendRequest(&retryConfig, r)
+		if err != nil {
+			return err
+		}
```

Comment:
```suggestion
		}
		defer res.Body.Close()
```

Reply:
separated this to a function so I can close the body there and then! I didn't want to use `defer` because that could potentially leave a lot of unclosed response body if we make a lot of concurrent graphql calls. Just want to close asap.